### PR TITLE
refactor auth provider

### PR DESCRIPTION
**Pull request by wanghaoran1988**
**What this PR does / why we need it**:
1. Abstract BearerTokenSource interface, `Token()` will generate a token, maybe cached or not, `Refresh()` will renew a token using refresh code and cached them, so next time invoke `Token()`, new token can be returned.
2. AuthProvider will generate a BearerTokenSource , so that client side can refresh tokens dynamically.
3. Keep the WrapTrasport() func in AuthProvider interface in case of the auth provider want to wrap something except the bearer token, like gce provider are clean the cache when 401.
4. And the BearerTokenSource param to rest config, so, if user can use it's own generated BeaerTokenSource instead of the one generate by authPrivider config.
5. And BearerTokenSource param to client config, so when use make rest request, they can decide whether to refresh the token dynamically
```

// BearerTokenSource will be used to get a token and refresh a new token
type BearerTokenSource interface {

	// Token returns a bearer token for HTTP requests.
	Token() (string, error)

	// This will try to refresh a new token and bust the cache.
	// Refresh can be called when the REST client experiences a
	// 401, which implies that the token is invalid or has expired.
	Refresh() (result bool, err error)
}

type AuthProvider interface {
	// WrapTransport allows the plugin to create a modified RoundTripper that
	// attaches authorization headers (or other info) to requests.
	WrapTransport(http.RoundTripper) http.RoundTripper
	// Login allows the plugin to initialize its configuration. It must not
	// require direct user interaction.
	Login() error

	// BearerTokenSource will return a BearerTokenSource from AuthProvider
	BearerTokenSource() BearerTokenSource
}
```

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:
```
Abstract a BearerTokenSource interface, so user can refresh the token dynamically.
```

### CODE REVIEW

**Comment by wanghaoran1988** [comment]
There is an issue when we bump only the BearerTokenSource, the --token options will be ignored again with this approach, should we also bump the basic auth and static token auth round ripper to the rest layer?

File: staging/src/k8s.io/client-go/rest/transport.go [hunk 1]
```diff
@@ -69,14 +71,15 @@ func (c *Config) TransportConfig() (*transport.Config, error) {
 			wt = func(rt http.RoundTripper) http.RoundTripper {
 				return provider.WrapTransport(previousWT(rt))
```
Comment (ericchiang) [inline staging/src/k8s.io/client-go/rest/transport.go]: Why are we still wrapping the transport?

File: staging/src/k8s.io/client-go/rest/client.go [hunk 2]
```diff
@@ -215,16 +219,30 @@ func createSerializers(config ContentConfig) (*Serializers, error) {
...
 //
 func (c *RESTClient) Verb(verb string) *Request {
 	backoff := c.createBackoffMgr()
-
+	var request *Request
 	if c.Client == nil {
-		return NewRequest(nil, verb, c.base, c.versionedAPIPath, c.contentConfig, c.serializers, backoff, c.Throttle)
+		request = NewRequest(nil, verb, c.base, c.versionedAPIPath, c.contentConfig, c.serializers, backoff, c.Throttle)
+	}
+	request = NewRequest(c.Client, verb, c.base, c.versionedAPIPath, c.contentConfig, c.serializers, backoff, c.Throttle)
```
Comment (ericchiang) [inline staging/src/k8s.io/client-go/rest/client.go]: This needs to be in an `else` block of the previous if statement.

File: staging/src/k8s.io/client-go/rest/client.go [hunk 3]
```diff
@@ -215,16 +219,30 @@ func createSerializers(config ContentConfig) (*Serializers, error) {
...
+	if c.BearerTokenSource != nil {
+		token, err := c.BearerTokenSource.Token()
+		if err != nil {
+			return request
+		}
+		request.SetHeader("Authorization", "Bearer "+token)
```
Comment (ericchiang) [inline staging/src/k8s.io/client-go/rest/client.go]: Worth adding a comment that bearer tokens in the kubeconfig override this later.

File: staging/src/k8s.io/client-go/rest/client.go [hunk 4]
```diff
@@ -215,16 +219,30 @@ func createSerializers(config ContentConfig) (*Serializers, error) {
 //  SelectorParam("labels", "area=staging").
 //  Timeout(10*time.Second).
 //  Do()
+// if resp.Status == http.StatusUnauthorized {
```
Comment (ericchiang) [inline staging/src/k8s.io/client-go/rest/client.go]: What's this comment for?

File: staging/src/k8s.io/client-go/rest/client.go [hunk 5]
```diff
@@ -91,7 +94,7 @@ type Serializers struct {
 // NewRESTClient creates a new RESTClient. This client performs generic REST functions
 // such as Get, Put, Post, and Delete on specified paths.  Codec controls encoding and
 // decoding of responses from the server.
-func NewRESTClient(baseURL *url.URL, versionedAPIPath string, config ContentConfig, maxQPS float32, maxBurst int, rateLimiter flowcontrol.RateLimiter, client *http.Client) (*RESTClient, error) {
+func NewRESTClient(baseURL *url.URL, versionedAPIPath string, config ContentConfig, maxQPS float32, maxBurst int, rateLimiter flowcontrol.RateLimiter, client *http.Client, bearerTokenSource BearerTokenSource) (*RESTClient, error) {
```
Comment (ericchiang) [inline staging/src/k8s.io/client-go/rest/client.go]: Probably best to not break the function signature. What's the code path from kubeconfig -> RESTClient. Can we just modify the RESTClient later?

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go [hunk 6]
```diff
@@ -124,10 +127,14 @@ func newGCPAuthProvider(_ string, gcpConfig map[string]string, persister restcli
...
-	return &conditionalTransport{&oauth2.Transport{Source: g.tokenSource, Base: rt}, g.persister}
+	return &cleanConfigTransport{rt, g.persister}
+}
+
+func (g *gcpAuthProvider) BearerTokenSource() restclient.BearerTokenSource {
+	return g.tokenSource
 }
 
-func (g *gcpAuthProvider) Login() error { return nil }
+func (g *gcpAuthProvider) Login() error { return errors.New("not yet implemented") }
```
Comment (ericchiang) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go]: nit: please avoid unrelated changes. It just makes the diff bigger.

**Review by ericchiang — COMMENTED** [review]
Biggest questions. Why are we keeping the WrapTransport auth method? Why are we modify the signature of NewRESTClient, can we make this change without that change?

edit on:

> There is an issue when we bump only the BearerTokenSource, the --token options will be ignored again with this approach, should we also bump the basic auth and static token auth round ripper to the rest layer?

I don't know. I think we want to make sure this PR doesn't get dragged out too much longer. I'm in favor of keeping this as small as possible, unless you have some concrete proposal for including the other auth stuff.

**Comment by wanghaoran1988** [comment]
> 3. Keep the WrapTrasport() func in AuthProvider interface in case of the auth provider want to wrap something except the bearer token, like gce provider are clean the cache when 401.

@ericchiang I was try to remove this, but found that gcp auth provider are try to wrap the some logic except the bearer token.https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go#L301

File: staging/src/k8s.io/client-go/rest/client.go (hunk 4 above)
Comment (wanghaoran1988) [inline staging/src/k8s.io/client-go/rest/client.go]: it's a doc example

File: staging/src/k8s.io/client-go/rest/client.go (hunk 3 above)
Comment (wanghaoran1988) [inline staging/src/k8s.io/client-go/rest/client.go]: @ericchiang The header from here will override the one from the bearer tokens in the kubeconfig as well as the one from --token cmd line, [code] here(https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/transport/round_trippers.go#L278), I will update here to align with the current logic, and add comments here.

File: staging/src/k8s.io/client-go/rest/client.go [hunk 7]
```diff
@@ -78,6 +78,10 @@ type RESTClient struct {
 
 	// Set specific behavior of the client.  If not set http.DefaultClient will be used.
 	Client *http.Client
+
+	// BearerTokenSource will be used to get a token and refresh a new token, it is generated from the auth provider
+	// or the one from rest.Config
+	BearerTokenSource BearerTokenSource
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/client.go]: I still don't think a token source fits well at the top level here. Authentication plugins have been provided as a custom client before. Why do we need two ways to feed that to a rest client now?

**Comment by liggitt** [comment]
Rather than plumb these changes throughout the rest client code, I'd rather limit this change to refactoring common function out of the existing token-based auth providers, and have the auth providers make use of that component. That avoids complicating the rest client builder code with multiple ways to specify auth wrappers, and preserves the power of an auth provider to wrap the transport in different ways.

It could also be proved out as a simple refactor of one of the existing providers without affecting any other packages:

1. a "token wrapper" component
2. a "token provider" interface
3. an implementation of the "token provider" interface

**Comment by ericchiang** [comment]
@liggitt main reason for this refactor was to get the retry logic capabilities without doing it in the transport.

E.g. the openstack auth plugin (https://github.com/kubernetes/kubernetes/pull/39587) currently does

```
// RoundTrip adds the bearer token into the request, and rotate it automatically
func (t *tokenRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
	req.Header.Set("Authorization", "Bearer "+t.token)
	resp, err := t.RoundTripper.RoundTrip(req)
	if err != nil || resp == nil {
		return resp, err
	}
	if resp.StatusCode != http.StatusUnauthorized {
		return resp, err
	}
	// will need to refresh the token, and redo the request
	if t.token, err = t.refreshFunc(); err != nil {
		return resp, err
	}
	if t.persister != nil {
		t.persister.Persist(map[string]string{"token": t.token})
	}
	newReq := cloneRequest(req)
	newReq.Header.Set("Authorization", "Bearer "+t.token)
	return t.RoundTripper.RoundTrip(newReq)
}
```

Maybe we could make the RESTClient retry on a 401 in certain situations? That way the auth client transport can refresh after it sees a 401 but not have to do the retry within the transport?

Or maybe we should just have the auth client transport just hit the 401 and not worry about a retry?

**Comment by anguslees** [comment]
> main reason for this refactor was to get the retry logic capabilities without doing it in the transport.

.. so what's actually _wrong_ with a roundtripper doing a retry?
or to phrase it differently: in the general case we need _some_ hook for an auth plugin to control a flow that involves multiple http requests, possibly involving the "real" http request - so if the retry code wasn't here it would have to be something similar somewhere nearby.  Is it just that we called this "transport" and the english word implies a slightly different interaction?

**Comment by ericchiang** [comment]
@anguslees 

http.RoundTripper's doc string says https://golang.org/pkg/net/http/#RoundTripper

        // RoundTrip should not attempt to interpret the response. In
        // particular, RoundTrip must return err == nil if it obtained
        // a response, regardless of the response's HTTP status code.
        // A non-nil err should be reserved for failure to obtain a
        // response. Similarly, RoundTrip should not attempt to
        // handle higher-level protocol details such as redirects,
        // authentication, or cookies.

Retrying also assumes the server doesn't consume any of the request body on a forbidden. So retrying within a RoundTripper makes a ton of assumptions that aren't guaranteed by the actual interfaces it deals with.

The REST client in Kubernetes already handles retries. I'd really prefer we do this the right way and not rely on behavior that's not guaranteed. A change to those interfaces would introduce really hard to debug error conditions.

**Comment by anguslees** [comment]
> http.RoundTripper's doc string says

Huh, look at that :)  I'd seen so many examples of bearer token / cookie handling in roundtrippers that I just thought that was an intended use case.
